scripts/protocol-dts-generator.ts

```diff
@@ -75,153 +66,197 @@ const emitDomain = (domain: P.Domain) => {
+    if (command.parameters) {
+        emitLine()
+        emitInterface(toCmdRequestName(command.name), command.parameters)
     }
 
-    const commandDef: ClientDef = {
-        name: command.name,
-        client: clientDef,
+    if (command.returns) {
+        emitLine()
+        emitInterface(toCmdResponseName(command.name), command.returns)
     }
+}
+
+const toEventPayloadName = (eventName: string) => `${toTitleCase(eventName)}Event`
 
-    return commandDef
-}
-
-const emitEvent = (event: P.Event, domain: string): ClientDef => {
-    const titleCase = toTitleCase(event.name)
-    const eventType = event.parameters ? `${domain}.${emitInterface(`${titleCase}Event`, event.parameters)}` : ''
-    const paramsDef = eventType ? [{name: 'params', $ref: eventType}] : []
-
-    const clientDef: P.FunctionType = {
-        type: 'function',
-        description: event.description,
-        name: `on`,
-        accepts: [
-            {
-                name: "event",
-                type: "string",
-                val: event.name,
-            },
-            {
-                name: "listener",
-                type: "lambda",
-                accepts: paramsDef
-            }
-        ],
+const emitEvent = (event: P.Event) => {
+    if (!event.parameters) {
+        return
     }
+    
+    emitLine()
+    emitDescription(event.description)
+    emitInterface(toEventPayloadName(event.name), event.parameters)
+}
 
-    const eventDef: ClientDef = {
-        name: event.name,
-        client: clientDef,
+const getEventMapping = (event: P.Event, domainName: string, modulePrefix: string): P.RefType & P.PropertyBaseType => {
+    // Use TS3.0+ tuples
+    const payloadType = event.parameters ?
+        `[${modulePrefix}.${domainName}.${toEventPayloadName(event.name)}]` :
+        '[]'
+
+    return {
+        // domain-prefixed name since it will be used outside of the module.
+        name: `${domainName}.${event.name}`,
+        // TODO(bckenny): does having a description help anywhere?
+        // description: event.description,
+        $ref: payloadType
     }
+}
+
+const isWeakInterface = (params: P.PropertyType[]): boolean => {
+    return params.every(p => !!p.optional)
+}
 
-    return eventDef
+const getCommandMapping = (command: P.Command, domainName: string, modulePrefix: string): P.ObjectType & P.PropertyBaseType => {
+    const prefix = `${modulePrefix}.${domainName}.`
+    // Use TS3.0+ tuples for paramsType
+    let requestType = '[]'
+    if (command.parameters) {
+        const optional = isWeakInterface(command.parameters) ? '?' : ''
+        requestType = '[' + prefix + toCmdRequestName(command.name) + optional + ']'
+    }
+    const responseType = command.returns ? prefix + toCmdResponseName(command.name) : 'void'
+
+    return {
+        type: 'object',
+        name: `${domainName}.${command.name}`,
+        // TODO(bckenny): does having a description help anywhere?
+        // description: command.description,
```

Comment:
what's the harm in having description?

Reply:
> what's the harm in having description?

no harm except file size and maybe ts parse time? The main thing is I couldn't find any place where the descriptions end up surfaced anyways since the mapping objects can't really be used directly, so it didn't seem worth it. I left them commented out so if it turns out we do want them you don't have to wade into the code again.

It seemed maybe better to take the command descriptions and somehow modify them to apply them to the `CommandRequest` and `CommandResponse` types, as those do show up in completions and intellisense, but I didn't see an easy way to modify the strings to apply the types. "The request parameters for a command that..." or something :)